### ISSUE TITLE
Any reason that Tensorflow.js does a much better job?

### ISSUE THREAD

**Issue by crearo** [rwightman/posenet-python#11]
Was curious on why the demo at https://storage.googleapis.com/tfjs-models/demos/posenet/camera.html does a much better job at detecting things than this repo -- even though it's a source clone as you say?
Note, I don't have metrics, and I might be terribly wrong, but that's my first impression on trying this out. I'm willing to contribute!

**Comment by rwightman** [comment]
I did most of my comparisons on the jpeg image set referenced in image_demo. Comparing the results at same resolutions and settings they seemed very close to the output when feeding the images to the JS model and analyzing the part outputs and probabilities. That being said I did those comparisons once, and haven't done a formal test script that can be run over time.

For the webcam there are more variables. I don't really know how the webcam is being controlled in the JS version. What the resolution is, whether the webcam's exposure/frame rate, etc is being controlled and set to more optimal values than in the defaults hidden behind the Python opencv webcam API. Differences in brightness and resolution could have a big impact.

Also, matching the various thresholds of the algorithm is important, the defaults may be different between the JS webcam demo and the Python one I threw together.

**Comment by crearo** [comment]
Ah alright. Let me try and experiment further. Will put up whatever I find!

**Comment by rwightman** [comment]
Sounds good. I'd check the threshold settings first, then resolution/aspect ratio, and then try to evaluate exposure/brightness.

For resolution, figure out what the size and aspect ratio of the camera output being fed into the model when running in JS is compared to Python. I think the JS impl tries to scale to 600x500ish? The python code may default to very different resolutions as asks for a 1MP resolution but really seems to get what it gets on different webcams, you could try requesting lower resolutions or rescaling the output to the same as the JS impl. The underlying NN is sensitive to scale and aspect. Saving a capture of images from each path and then feeding those back into the image input mode would be useful for comparing more precisely. 

Brightness of images will be a bit more challenging, a level of brightness difference that impacts the algorithm may not be as evident to you. If default frame rates differ ie one is 30fps and the other 60, there will likely be a default exposure will be capped at 1/30s vs 1/60 and you can see variation within that if there is an auto-exposure or auto-gain algorithm active that is trying to modulate the exposure or post-gain to keep a consistent output appearance in different lighting.

**Comment by crearo** [comment]
Alright, right away here's what I got:

1. Resolution
Yes, the `js` one uses 600x500 (that's an odd resolution tbh). I changed the default to 640x480 and it magically just does much better.

2. Scale Factor
The `js` one was set to `0.5` by default, but I see you using `0.7`. And again, changing that made a whole lot of difference! 

Just the two above changed quite a bit.

3. Filters
I work at a thermal camera company and so before changing brightness/contrast, I straight up tested the feed on my thermal camera. It worked pretty well considering the model wasn't trained for thermal output! That tells me filtering the video won't drastically change the output. So I took 10 images of myself in visible light and increased the contrast and lowered the brightness. From my understanding of how posenet works - it works on gradients much like HoG, bringing out the edges in the video helps. Which it slightly did.

Honestly, with just these changes I'm pretty happy to go ahead and possibly train it on thermal now at the moment. 
Also, I've made a small PR at #12.

**Comment by gustavz** [comment]
@crearo you are mentioning training here

> Honestly, with just these changes I'm pretty happy to go ahead and possibly **train** it on thermal now at the moment.

Did you build up a pipeline to train the model? 
I would be very interesting in getting to know how to train the posenet, as it is not really straight forward as with other tensorflow models.
